libyul/optimiser/SSATransform.cpp

```diff
@@ -40,24 +42,57 @@ void SSATransform::operator()(Identifier& _identifier)
 
 void SSATransform::operator()(ForLoop& _for)
 {
-	// This will clear the current value in case of a reassignment inside the
-	// init part, although the new variable would still be in scope inside the whole loop.
-	// This small inefficiency is fine if we move the pre part of all for loops out
-	// of the for loop.
-	(*this)(_for.pre);
+	// Avoid dealing with complicated scoping rule
+	assertThrow(_for.pre.statements.empty(), OptimizerException, "");
+
+	// prepend statements in _a to the beginning of _b, also empties _a
+	auto prependStatements = [](vector<Statement>& _a, vector<Statement>& _b)
+	{
+		std::move(_b.begin(), _b.end(), std::back_inserter(_a));
+		swap(_a, _b);
+		_a.clear();
+	};
+
+	// If existing variables get re-assigned in loop's body block or post block,
+	// we need to declare new variables to hold its value at the location where
+	// control flow joins, i.e beginning of post block and beginning of body block.
 
 	Assignments assignments;
 	assignments(_for.body);
 	assignments(_for.post);
-	for (auto const& var: assignments.names())
-		m_currentVariableValues.erase(var);
 
+	// clear values of variable re-assigned in _for.body and _for.post
+	auto clearVariableValues = [&]()
+	{
+		for (YulString varName: assignments.names())
+			m_currentVariableValues.erase(varName);
+	};
+
+	// stores variables declared outside the loop but re-assigned in post/body block.
+	vector<YulString> varsToDeclare;
+	for (YulString varName: assignments.names())
+		if (m_currentVariableValues.count(varName))
+			varsToDeclare.push_back(varName);
+
+	clearVariableValues();
 	visit(*_for.condition);
+
+	// temporarily holds declarations to be prepend to the block
+	vector<Statement> decls;
+
+	for (YulString varName: varsToDeclare)
+		decls.emplace_back(assignVarToSSAVar(varName, _for.location));
 	(*this)(_for.body);
+	prependStatements(decls, _for.body.statements);
+	clearVariableValues();
+
+	for (YulString varName: varsToDeclare)
+		decls.emplace_back(assignVarToSSAVar(varName, _for.location));
 	(*this)(_for.post);
+	prependStatements(decls, _for.post.statements);
```

Comment:
Maybe
```suggestion
	_for.post.statements = std::move(_for.post.statements) + std::move(decls);
```
would be easier to read?

Reply:
Yes! I didn't know we have `operator+` defined for vector in libdevcore.